File: src/main/java/com/yupi/template/service/NanoBananaService.java
package com.yupi.template.service;

import com.google.genai.Client;
import com.google.genai.types.GenerateContentConfig;
import com.google.genai.types.GenerateContentResponse;
import com.google.genai.types.ImageConfig;
import com.google.genai.types.Part;
import com.yupi.template.config.NanoBananaConfig;
import com.yupi.template.model.enums.ImageMethodEnum;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;

import jakarta.annotation.Resource;

import static com.yupi.template.constant.ArticleConstant.PICSUM_URL_TEMPLATE;

/**
 * Nano Banana (Gemini 原生图片生成) 服务
 * 使用 Gemini 2.5 Flash Image 或 Gemini 3 Pro Image 模型生成图片
 *
 * @author <a href="https://codefather.cn">编程导航学习圈</a>
 */
@Service
@Slf4j
public class NanoBananaService implements ImageSearchService {

    @Resource
    private NanoBananaConfig nanoBananaConfig;

    @Resource
    private CosService cosService;

    @Override
    public String searchImage(String keywords) {
        // 对于 Nano Banana，keywords 就是生图 prompt
        return generateImage(keywords);
    }

    /**
     * 根据提示词生成图片
     *
     * @param prompt 生图提示词
     * @return 图片 URL，生成失败返回 null
     */
    public String generateImage(String prompt) {
        try {
            // 使用 Builder 显式设置 API Key
            Client genaiClient = Client.builder()
                    .apiKey(nanoBananaConfig.getApiKey())
                    .build();
            
            try {
                // 构建图片配置
                ImageConfig.Builder imageConfigBuilder = ImageConfig.builder()
                        .aspectRatio(nanoBananaConfig.getAspectRatio());

                // Gemini 3 Pro Image 支持更高分辨率
                String model = nanoBananaConfig.getModel();
                if (model != null && model.contains("gemini-3-pro")) {
                    imageConfigBuilder.imageSize(nanoBananaConfig.getImageSize());
                }

                // 构建生成配置
                GenerateContentConfig config = GenerateContentConfig.builder()
                        .responseModalities("TEXT", "IMAGE")
                        .imageConfig(imageConfigBuilder.build())
                        .build();

                log.info("Nano Banana 开始生成图片, model={}, prompt={}", model, prompt);

                // 调用 Gemini API 生成图片
                GenerateContentResponse response = genaiClient.models.generateContent(
                        model != null ? model : "gemini-2.5-flash-image",
                        prompt,
                        config);

                // 从响应中提取图片数据
                if (response.parts() != null) {
                    for (Part part : response.parts()) {
                        if (part.inlineData().isPresent()) {
                            var blob = part.inlineData().get();
                            if (blob.data().isPresent()) {
                                byte[] imageBytes = blob.data().get();
                                String mimeType = blob.mimeType().orElse("image/png");
                                
                                log.info("Nano Banana 图片生成成功, size={} bytes, mimeType={}", 
                                        imageBytes.length, mimeType);
                                
                                // 上传图片到 COS 并返回 URL
                                return uploadImageToCos(imageBytes, mimeType);
                            }
                        }
                    }
                }

                log.warn("Nano Banana 未生成图片, prompt={}", prompt);
                return null;

            } finally {
                genaiClient.close();
            }
        } catch (Exception e) {
            log.error("Nano Banana 生成图片异常, prompt={}", prompt, e);
            return null;
        }
    }

    /**
     * 上传图片字节数据到 COS
     *
     * @param imageBytes 图片字节数据
     * @param mimeType   图片 MIME 类型
     * @return COS 图片 URL
     */
    private String uploadImageToCos(byte[] imageBytes, String mimeType) {
        try {
            // 生成临时文件名
            String extension = mimeType.contains("jpeg") || mimeType.contains("jpg") ? ".jpg" : ".png";
            String fileName = "nano-banana/" + System.currentTimeMillis() + "_" + 
                    java.util.UUID.randomUUID().toString().substring(0, 8) + extension;

            // 使用 CosService 上传（需要先转换为 URL 或直接上传字节）
            // 由于 CosService 目前只支持 URL 下载上传，这里直接使用 base64 data URL
            // 或者扩展 CosService 支持字节上传
            
            // 临时方案：将图片转为 base64 data URL（前端可直接使用）
            String base64 = java.util.Base64.getEncoder().encodeToString(imageBytes);
            String dataUrl = "data:" + mimeType + ";base64," + base64;
            
            log.info("Nano Banana 图片已生成 Data URL, length={}", dataUrl.length());
            return dataUrl;
            
        } catch (Exception e) {
            log.error("上传 Nano Banana 图片失败", e);
            return null;
        }
    }

    @Override
    public ImageMethodEnum getMethod() {
        return ImageMethodEnum.NANO_BANANA;
    }

    @Override
    public String getFallbackImage(int position) {
        return String.format(PICSUM_URL_TEMPLATE, position);
    }
}

File: python-backend/app/services/nano_banana_service.py
"""Nano Banana (Gemini 原生图片生成) 服务（第 5 期新增）"""

import logging
from typing import Optional
from google import genai
from google.genai import types

from app.config import settings
from app.constants.article import ArticleConstant
from app.models.enums import ImageMethodEnum
from app.services.image_search_service import ImageSearchService
from app.schemas.image import ImageData, ImageRequest

logger = logging.getLogger(__name__)


class NanoBananaService(ImageSearchService):
    """Nano Banana (Gemini 原生图片生成) 服务"""
    
    def __init__(self):
        self.api_key = settings.nano_banana_api_key
        self.model = settings.nano_banana_model
        self.aspect_ratio = settings.nano_banana_aspect_ratio
        self.image_size = settings.nano_banana_image_size
        # 初始化 Gemini 客户端
        self.client = genai.Client(api_key=self.api_key)
    
    async def search_image(self, keywords: str) -> Optional[str]:
        """此方法已废弃，请使用 get_image_data()"""
        return None
    
    async def get_image_data(self, request: ImageRequest) -> Optional[ImageData]:
        """获取图片数据"""
        prompt = request.get_effective_param(True)
        return await self.generate_image_data(prompt)
    
    async def generate_image_data(self, prompt: str) -> Optional[ImageData]:
        """
        根据提示词生成图片数据
        
        Args:
            prompt: 生图提示词
            
        Returns:
            ImageData 包含图片字节数据，生成失败返回 None
        """
        try:
            # 构建图片配置
            image_config_params = {"aspect_ratio": self.aspect_ratio}
            
            # Gemini 3 Pro Image 支持更高分辨率
            if self.model and "gemini-3-pro" in self.model:
                image_config_params["image_size"] = self.image_size
            
            # 使用 types.ImageConfig 构建配置对象
            image_config = types.ImageConfig(**image_config_params)
            
            # 构建生成配置
            config = types.GenerateContentConfig(
                response_modalities=["TEXT", "IMAGE"],
                image_config=image_config
            )
            
            logger.info(f"Nano Banana 开始生成图片, model={self.model}, prompt={prompt}")
            
            # 调用 Gemini API 生成图片
            response = self.client.models.generate_content(
                model=self.model or "gemini-2.5-flash-image",
                contents=prompt,
                config=config
            )
            
            # 从响应中提取图片数据
            if response.candidates and response.candidates[0].content.parts:
                for part in response.candidates[0].content.parts:
                    if part.inline_data:
                        image_bytes = part.inline_data.data
                        mime_type = part.inline_data.mime_type or "image/png"
                        
                        logger.info(
                            f"Nano Banana 图片生成成功, "
                            f"size={len(image_bytes)} bytes, mimeType={mime_type}"
                        )
                        
                        return ImageData.from_bytes(image_bytes, mime_type)
            
            logger.error("Nano Banana 响应中未找到图片数据")
            return None
        except Exception as e:
            logger.error(f"Nano Banana 生成图片异常: {e}")
            return None
    
    def get_method(self) -> ImageMethodEnum:
        """获取图片服务类型"""
        return ImageMethodEnum.NANO_BANANA
    
    def get_fallback_image(self, position: int) -> str:
        """获取降级图片"""
        return ArticleConstant.PICSUM_URL_TEMPLATE.format(position)
